### ISSUE TITLE
`brace_linter()` could allow skipping braces in function factories

### ISSUE THREAD

**Issue by AshesITR** [jimhester/lintr#1807]
``` r
lintr::lint(text = "function() function(x) {\n  x\n}")
#> <text>:1:1: style: [brace_linter] Any function spanning multiple lines should use curly braces.
#> function() function(x) {
#> ^~~~~~~~~~~~~~~~~~~~~~~~
```

<sup>Created on 2022-12-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This could optionally be allowed in addition to
```r
# always allowed
function() {
  function(x) {
    x
  }
}

# could optionally be allowed
function() function(x) {
  x
}
```

**Comment by MichaelChirico** [comment]
WDYT about making it the same option as allowing e.g.

```
function(x) AWrapper(
  y,
  x
)
```

?

**Comment by AshesITR** [comment]
Wouldn't that disable that particular lint completely?

**Comment by MichaelChirico** [comment]
No, there's still

```r
function(x)
  AWrapper(y, x)

function(x)
  AWrapper(
    y,
    x
  )
```

**Comment by AshesITR** [comment]
Ah okay. Any ideas for an option name? `allow_simple_multiline = FALSE` is what comes to my mind, but I'm not really happy with it.

**Comment by IndrajeetPatil** [comment]
But omitting braces here reduces readability, since the reader now needs to mentally add those braces back in. 

```r
function() function(x) {
  x
}
```

Why should we make an exception for such an omission?  That is, why is it so important here to leave out braces?

**Comment by AshesITR** [comment]
Because you need two additional lines and an additional level of indent.
Not arguing that this should be default, but if you define a bunch of function factories, you end up with a lot of essentially empty lines and reduced available line width for the actual implementations.

**Comment by MichaelChirico** [comment]
Agree it can be useful, and warrants an argument, but also that it should not be the default choice.

**Comment by IndrajeetPatil** [comment]
Okay, thanks for the explaining. Makes sense.

As for the parameter name, how about `curly_braces_required` (default: `TRUE`)?

**Comment by MichaelChirico** [comment]
is `strict=TRUE` too glib?

**Comment by AshesITR** [comment]
`strict` sounds okay, while not very specific. `curly_braces_required` would suggest that the other examples mentioned by @MichaelChirico don't lint, which they always should.
